 ## Data Cleaning

In [1]:
#!pip install --upgrade openpyxl

In [2]:
#importing libraries and installing packages needed

import pandas as pd

C:\Users\hello\anaconda3\Lib\site-packages\pandas\core\arrays\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


In [3]:
# Load the two worksheets
df1 = pd.read_excel(
    "challenge dataset.xlsx", 
    sheet_name='Candidat au don 2019 (avec age)', 
    dtype={
        'ÉLIGIBILITÉ AU DON.': str,
        'A-t-il_(elle)_déjà_donné_le_sang_': str
    })
df2 = pd.read_excel("challenge dataset.xlsx", sheet_name='Donneurs 2019')

#first five columns of 'Candidat au don 2019 (avec age)' worksheet
df1.head()

ID  Age Horodateur Niveau_scolaire Genre_  Taille_  Poids  \
0  DONOR_1842   24    46:44.9   Universitaire  Homme      NaN    NaN   
1  DONOR_1843   23    44:19.0   Universitaire  Homme      NaN    NaN   
2  DONOR_1844   41    46:54.3           Aucun  Homme      NaN    NaN   
3  DONOR_1845   24    29:29.7   Universitaire  Homme      NaN    NaN   
4  DONOR_1846   30    47:47.4      Secondaire  Homme      NaN    NaN   

  Situation_Matrimoniale_(SM)   Profession_ Arrondissement_de_résidence_  ...  \
0                 Célibataire  Etudiant (e)                     Douala 3  ...   
1                 Célibataire  Sans Emplois                        R A S  ...   
2                   Marié (e)     Tailleur                      Douala 1  ...   
3                 Célibataire      Etudiant                     Douala 5  ...   
4                 Célibataire     Militaire        Douala (Non précisé )  ...   

  Raison_de_non-eligibilité_totale__[Porteur(HIV,hbs,hcv)]  \
0                                                NaN         
1                                                Non         
2                                                Non         
3                                                NaN         
4                                                NaN         

  Raison_de_non-eligibilité_totale__[Opéré]  \
0                                       NaN   
1                                       Oui   
2                                       Non   
3                                       NaN   
4                                       NaN   

  Raison_de_non-eligibilité_totale__[Drepanocytaire]  \
0                                                NaN   
1                                                Non   
2                                                Non   
3                                                NaN   
4                                                NaN   

  Raison_de_non-eligibilité_totale__[Diabétique]  \
0                                            NaN   
1                                            Non   
2                                            Non   
3                                            NaN   
4                                            NaN   

  Raison_de_non-eligibilité_totale__[Hypertendus]  \
0                                             NaN   
1                                             Non   
2                                             Non   
3                                             NaN   
4                                             NaN   

  Raison_de_non-eligibilité_totale__[Asthmatiques]  \
0                                              NaN   
1                                              Non   
2                                              Non   
3                                              NaN   
4                                              NaN   

  Raison_de_non-eligibilité_totale__[Cardiaque]  \
0                                           NaN   
1                                           Non   
2                                           Non   
3                                           NaN   
4                                           NaN   

  Raison_de_non-eligibilité_totale__[Tatoué]  \
0                                        NaN   
1                                        Non   
2                                        Non   
3                                        NaN   
4                                        NaN   

  Raison_de_non-eligibilité_totale__[Scarifié] Si_autres_raison_préciser_  
0                                          NaN                        NaN  
1                                          Non                        NaN  
2                                          Non        Aucune information   
3                                          NaN                        NaN  
4                                          NaN                        NaN  

[5 rows x 40 columns]

In [4]:
#first five coumns of 'Donneur 2019' worksheet
df2.head()

Horodateur Sexe Age  Type de donation   \
0 2020-07-16 19:02:33.994    M   26                 F   
1 2020-07-16 19:05:51.521    M   24                 F   
2 2020-07-17 04:52:09.054    M   41                 F   
3 2020-07-17 04:56:03.200    M   20                 F   
4 2020-07-17 05:49:40.730    M   34                 F   

  Groupe Sanguin ABO / Rhesus                       Phenotype   
0                           B+  RH-2, RH+4, RH-3, RH+5, -kell1  
1                           O+  RH-2, RH+4, RH-3, RH+5, -kell1  
2                           A+  RH+2, RH+4, RH-3, RH+5, -kell1  
3                           O+  RH-2, RH+4, RH-3, RH+5, -kell1  
4                           O+  RH-2, RH+4, RH+3, RH+5, -kell1

Moving on with cleaning one dataframe at a time. 

---

### Cleaning 'Candidat au don 2019 (avec age)' Sheet 

In [5]:
df1

ID  Age Horodateur Niveau_scolaire Genre_  Taille_  Poids  \
0     DONOR_1842   24    46:44.9   Universitaire  Homme      NaN    NaN   
1     DONOR_1843   23    44:19.0   Universitaire  Homme      NaN    NaN   
2     DONOR_1844   41    46:54.3           Aucun  Homme      NaN    NaN   
3     DONOR_1845   24    29:29.7   Universitaire  Homme      NaN    NaN   
4     DONOR_1846   30    47:47.4      Secondaire  Homme      NaN    NaN   
...          ...  ...        ...             ...    ...      ...    ...   
1841  DONOR_3683   26    31:07.0      Secondaire  Homme      NaN    NaN   
1842  DONOR_3684   23    29:02.4      Secondaire  Homme      NaN    NaN   
1843  DONOR_3685   30    33:42.3   Universitaire  Homme      NaN    NaN   
1844  DONOR_3686   28    02:54.5     Pas Précisé  Femme      NaN   96.0   
1845  DONOR_3687   29    56:38.9     Pas Précisé  Homme      NaN    NaN   

     Situation_Matrimoniale_(SM)           Profession_  \
0                    Célibataire          Etudiant (e)   
1                    Célibataire          Sans Emplois   
2                      Marié (e)             Tailleur    
3                    Célibataire              Etudiant   
4                    Célibataire             Militaire   
...                          ...                   ...   
1841                 Célibataire                 Élève   
1842                 Célibataire                Élève    
1843                 Célibataire  Technicien superieur   
1844                 Célibataire           Estheticien   
1845                 Célibataire           Électricien   

     Arrondissement_de_résidence_  ...  \
0                        Douala 3  ...   
1                           R A S  ...   
2                        Douala 1  ...   
3                        Douala 5  ...   
4           Douala (Non précisé )  ...   
...                           ...  ...   
1841                     Douala 3  ...   
1842        Douala (Non précisé )  ...   
1843        Douala (Non précisé )  ...   
1844        Douala (Non précisé )  ...   
1845                     Douala 3  ...   

     Raison_de_non-eligibilité_totale__[Porteur(HIV,hbs,hcv)]  \
0                                                   NaN         
1                                                   Non         
2                                                   Non         
3                                                   NaN         
4                                                   NaN         
...                                                 ...         
1841                                                NaN         
1842                                                NaN         
1843                                                NaN         
1844                                                NaN         
1845                                                NaN         

     Raison_de_non-eligibilité_totale__[Opéré]  \
0                                          NaN   
1                                          Oui   
2                                          Non   
3                                          NaN   
4                                          NaN   
...                                        ...   
1841                                       NaN   
1842                                       NaN   
1843                                       NaN   
1844                                       NaN   
1845                                       NaN   

     Raison_de_non-eligibilité_totale__[Drepanocytaire]  \
0                                                   NaN   
1                                                   Non   
2                                                   Non   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
1841                                                NaN   
1842                                                NaN  

In [6]:
# Display the column names
print(df1.columns.tolist())

['ID', 'Age', 'Horodateur', 'Niveau_scolaire', 'Genre_', 'Taille_', 'Poids', 'Situation_Matrimoniale_(SM)', 'Profession_', 'Arrondissement_de_résidence_', 'Quartier_de_Résidence_', 'Nationalité_', 'Religion_', 'A-t-il_(elle)_déjà_donné_le_sang_', 'Si_oui_preciser_la_date_du_dernier_don._', 'Taux_d’hémoglobine_', 'ÉLIGIBILITÉ_AU_DON.', 'Raison_indisponibilité__[Est_sous_anti-biothérapie__]', 'Raison_indisponibilité__[Taux_d’hémoglobine_bas_]', 'Raison_indisponibilité__[date_de_dernier_Don_<_3_mois_]', 'Raison_indisponibilité__[IST_récente_(Exclu_VIH,_Hbs,_Hcv)]', 'Date_de_dernières_règles_(DDR)__', 'Raison_de_l’indisponibilité_de_la_femme_[La_DDR_est_mauvais_si_<14_jour_avant_le_don]', 'Raison_de_l’indisponibilité_de_la_femme_[Allaitement_]', 'Raison_de_l’indisponibilité_de_la_femme_[A_accoucher_ces_6_derniers_mois__]', 'Raison_de_l’indisponibilité_de_la_femme_[Interruption_de_grossesse__ces_06_derniers_mois]', 'Raison_de_l’indisponibilité_de_la_femme_[est_enceinte_]', 'Autre_raisons,__

In [7]:
df1.shape

(1846, 40)

In [8]:
# Count NaN values per column
nan_count = df1.isna().sum()
print(nan_count)

ID                                                                                              0
Age                                                                                             0
Horodateur                                                                                      0
Niveau_scolaire                                                                                 0
Genre_                                                                                          0
Taille_                                                                                      1786
Poids                                                                                        1742
Situation_Matrimoniale_(SM)                                                                     0
Profession_                                                                                     0
Arrondissement_de_résidence_                                                                    0
Quartier_de_Résidenc

Given that the number of rows in the datset is 1846 and some columns have 95.77% of missing data , we shall proceed to droping them 

In [9]:
# Drop rows with any NaN values
clean_df1 = df1.drop(columns=[ 'Taille_','Poids','Taux_d’hémoglobine_', 'Si_oui_preciser_la_date_du_dernier_don._','Raison_indisponibilité__[Est_sous_anti-biothérapie__]', 'Raison_indisponibilité__[Taux_d’hémoglobine_bas_]', 'Raison_indisponibilité__[date_de_dernier_Don_<_3_mois_]', 'Raison_indisponibilité__[IST_récente_(Exclu_VIH,_Hbs,_Hcv)]', 'Date_de_dernières_règles_(DDR)__', 'Raison_de_l’indisponibilité_de_la_femme_[La_DDR_est_mauvais_si_<14_jour_avant_le_don]', 'Raison_de_l’indisponibilité_de_la_femme_[Allaitement_]', 'Raison_de_l’indisponibilité_de_la_femme_[A_accoucher_ces_6_derniers_mois__]', 'Raison_de_l’indisponibilité_de_la_femme_[Interruption_de_grossesse__ces_06_derniers_mois]', 'Raison_de_l’indisponibilité_de_la_femme_[est_enceinte_]', 'Autre_raisons,__preciser_', 'Sélectionner_"ok"_pour_envoyer_', 'Raison_de_non-eligibilité_totale__[Antécédent_de_transfusion]', 'Raison_de_non-eligibilité_totale__[Porteur(HIV,hbs,hcv)]', 'Raison_de_non-eligibilité_totale__[Opéré]', 'Raison_de_non-eligibilité_totale__[Drepanocytaire]', 'Raison_de_non-eligibilité_totale__[Diabétique]', 'Raison_de_non-eligibilité_totale__[Hypertendus]', 'Raison_de_non-eligibilité_totale__[Asthmatiques]', 'Raison_de_non-eligibilité_totale__[Cardiaque]', 'Raison_de_non-eligibilité_totale__[Tatoué]', 'Raison_de_non-eligibilité_totale__[Scarifié]', 'Si_autres_raison_préciser_'])
clean_df1

ID  Age Horodateur Niveau_scolaire Genre_  \
0     DONOR_1842   24    46:44.9   Universitaire  Homme   
1     DONOR_1843   23    44:19.0   Universitaire  Homme   
2     DONOR_1844   41    46:54.3           Aucun  Homme   
3     DONOR_1845   24    29:29.7   Universitaire  Homme   
4     DONOR_1846   30    47:47.4      Secondaire  Homme   
...          ...  ...        ...             ...    ...   
1841  DONOR_3683   26    31:07.0      Secondaire  Homme   
1842  DONOR_3684   23    29:02.4      Secondaire  Homme   
1843  DONOR_3685   30    33:42.3   Universitaire  Homme   
1844  DONOR_3686   28    02:54.5     Pas Précisé  Femme   
1845  DONOR_3687   29    56:38.9     Pas Précisé  Homme   

     Situation_Matrimoniale_(SM)           Profession_  \
0                    Célibataire          Etudiant (e)   
1                    Célibataire          Sans Emplois   
2                      Marié (e)             Tailleur    
3                    Célibataire              Etudiant   
4                    Célibataire             Militaire   
...                          ...                   ...   
1841                 Célibataire                 Élève   
1842                 Célibataire                Élève    
1843                 Célibataire  Technicien superieur   
1844                 Célibataire           Estheticien   
1845                 Célibataire           Électricien   

     Arrondissement_de_résidence_ Quartier_de_Résidence_  Nationalité_  \
0                        Douala 3                  Yassa  Camerounaise   
1                           R A S                  R A S         R A S   
2                        Douala 1               NEW BELL  Camerounaise   
3                        Douala 5           Ange raphael  Camerounaise   
4           Douala (Non précisé )                  Beedi  Camerounaise   
...                           ...                    ...           ...   
1841                     Douala 3               Ndogbong  Camerounaise   
1842        Douala (Non précisé )                   Pk14  Camerounaise   
1843        Douala (Non précisé )      Cite des palmiers  Camerounaise   
1844        Douala (Non précisé )          Douala douala  Camerounaise   
1845                     Douala 3             PK16 NSAPE  Camerounaise   

                  Religion_ A-t-il_(elle)_déjà_donné_le_sang_  \
0     Chretien (Catholique)                               Non   
1     Chretien (Catholique)                               Oui   
2               Pas Précisé                               Non   
3                  Musulman                               Non   
4                Adventiste                               Oui   
...                     ...                               ...   
1841   Chrétien non précisé                               Non   
1842            Pas Précisé                               Oui   
1843               Musulman                               Oui   
1844            Pas Précisé                               Non   
1845  Chretien (Catholique)                               Non   

              ÉLIGIBILITÉ_AU_DON.  
0                        Eligible  
1     Définitivement non-eligible  
2     Définitivement non-eligible  
3                        Eligible  
4                        Eligible  
...                           ...  
1841                     Eligible  
1842                     Eligible  
1843                     Eligible  
1844                     Eligible  
1845                     Eligible  

[1846 rows x 13 columns]

In [10]:
print(clean_df1['ÉLIGIBILITÉ_AU_DON.'].unique())

['Eligible' 'Définitivement non-eligible' 'Temporairement Non-eligible']


In [11]:
# Display the first few rows of the eligibility column
print(clean_df1['ÉLIGIBILITÉ_AU_DON.'].head(20))  # Show first 20 entries

0                        Eligible
1     Définitivement non-eligible
2     Définitivement non-eligible
3                        Eligible
4                        Eligible
5                        Eligible
6                        Eligible
7                        Eligible
8                        Eligible
9     Temporairement Non-eligible
10                       Eligible
11                       Eligible
12                       Eligible
13                       Eligible
14    Temporairement Non-eligible
15                       Eligible
16                       Eligible
17    Temporairement Non-eligible
18                       Eligible
19                       Eligible
Name: ÉLIGIBILITÉ_AU_DON., dtype: object


In [12]:
#dftest = pd.read_excel("Updated Challenge dataset.xlsx", sheet_name='Candidat au don 2019 (avec age)', dtype={'ÉLIGIBILITÉ_AU_DON.': str})

In [13]:
#dftest['ÉLIGIBILITÉ_AU_DON.']

In [15]:
# ensuring consistency within the columns
clean_df1['Nationalité_'] = clean_df1['Nationalité_'].replace({
    'R A S': 'Camerounaise',
    'Ras': 'Camerounaise'
})
clean_df1['Profession_'] = clean_df1['Profession_'].replace({
    'R A S': 'Pas Précisé',
    'Ras': 'Pas Précisé'
})
clean_df1['Arrondissement_de_résidence_'] = clean_df1['Arrondissement_de_résidence_'].replace({
    'R A S': 'Pas Précisé',
    'Ras': 'Pas Précisé'
})
clean_df1['Quartier_de_Résidence_'] = clean_df1['Quartier_de_Résidence_'].replace({
    'R A S': 'Pas Précisé',
    'Ras': 'Pas Précisé'
})



In [16]:
# Handle Niveau_scolaire column by mapping categories to numerical values
level_mapping = {
    'Pas Précisé': -1,
    'Aucun':0,
    'Primaire': 1,
    'Secondaire': 2,
    'Universitaire': 3,
}

clean_df1['Niveau_scolaire'] = clean_df1['Niveau_scolaire'].replace(level_mapping)

# Check the changes
print(clean_df1[['Religion_', 'Nationalité_', 'Profession_', 'Niveau_scolaire']].head(20))

                 Religion_  Nationalité_       Profession_  Niveau_scolaire
0    Chretien (Catholique)  Camerounaise      Etudiant (e)                3
1    Chretien (Catholique)  Camerounaise      Sans Emplois                3
2              Pas Précisé  Camerounaise         Tailleur                 0
3                 Musulman  Camerounaise          Etudiant                3
4               Adventiste  Camerounaise         Militaire                2
5                 Musulman  Camerounaise      Etudiant (e)                3
6              Pas Précisé  Camerounaise      Sans Emplois                1
7                 Musulman  Camerounaise         Bijoutier               -1
8                 Musulman  Camerounaise         Couturier                2
9                 Musulman  Camerounaise       Jeune cadre                3
10                Musulman  Camerounaise    Commerçant (e)                3
11             Pas Précisé  Camerounaise       Mototaximan                2
12          

C:\Users\hello\AppData\Local\Temp\ipykernel_9248\1320290316.py:10: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  clean_df1['Niveau_scolaire'] = clean_df1['Niveau_scolaire'].replace(level_mapping)


In [17]:
clean_df1.head()

ID  Age Horodateur  Niveau_scolaire Genre_  \
0  DONOR_1842   24    46:44.9                3  Homme   
1  DONOR_1843   23    44:19.0                3  Homme   
2  DONOR_1844   41    46:54.3                0  Homme   
3  DONOR_1845   24    29:29.7                3  Homme   
4  DONOR_1846   30    47:47.4                2  Homme   

  Situation_Matrimoniale_(SM)   Profession_ Arrondissement_de_résidence_  \
0                 Célibataire  Etudiant (e)                     Douala 3   
1                 Célibataire  Sans Emplois                  Pas Précisé   
2                   Marié (e)     Tailleur                      Douala 1   
3                 Célibataire      Etudiant                     Douala 5   
4                 Célibataire     Militaire        Douala (Non précisé )   

  Quartier_de_Résidence_  Nationalité_              Religion_  \
0                  Yassa  Camerounaise  Chretien (Catholique)   
1                  R A S  Camerounaise  Chretien (Catholique)   
2               NEW BELL  Camerounaise            Pas Précisé   
3           Ange raphael  Camerounaise               Musulman   
4                  Beedi  Camerounaise             Adventiste   

  A-t-il_(elle)_déjà_donné_le_sang_          ÉLIGIBILITÉ_AU_DON.  
0                               Non                     Eligible  
1                               Oui  Définitivement non-eligible  
2                               Non  Définitivement non-eligible  
3                               Non                     Eligible  
4                               Oui                     Eligible

In [18]:
# Label Encoding for categorical columns
from sklearn.preprocessing import LabelEncoder

# List of columns to encode
columns_to_encode = [
    'Genre_',
    'Situation_Matrimoniale_(SM)',
    'A-t-il_(elle)_déjà_donné_le_sang_',
    'ÉLIGIBILITÉ_AU_DON.'
]

# Initialize the LabelEncoder
label_encoder = LabelEncoder()

# Apply label encoding to specified columns
for column in columns_to_encode:
    clean_df1[column] = label_encoder.fit_transform(clean_df1[column].astype(str))

# Display the updated DataFrame
print("\nDataFrame after Label Encoding:")
print(clean_df1[columns_to_encode].head())


DataFrame after Label Encoding:
   Genre_  Situation_Matrimoniale_(SM)  A-t-il_(elle)_déjà_donné_le_sang_  \
0       1                            0                                  0   
1       1                            0                                  1   
2       1                            2                                  0   
3       1                            0                                  0   
4       1                            0                                  1   

   ÉLIGIBILITÉ_AU_DON.  
0                    1  
1                    0  
2                    0  
3                    1  
4                    1  


In [ ]:
clean_df1